lightning-block-sync/src/lib.rs

```diff
@@ -0,0 +1,1047 @@
+	chain_notifier.sync_listener(new_header, &old_header, &mut chain_poller, chain_listener).await.unwrap();
+}
+
+/// Unbounded cache of header data keyed by block hash.
+pub(crate) type HeaderCache = std::collections::HashMap<BlockHash, ValidatedBlockHeader>;
+
+/// Keep the chain that a chain listener knows about up-to-date with the best chain from any of the
+/// given block_sources.
+///
+/// This implements a pretty bare-bones SPV client, checking all relevant commitments and finding
+/// the heaviest chain, but not storing the full header chain, leading to some important
+/// limitations.
+///
+/// TODO: Update comment to reflect this is now the responsibility of chain_poller.
+/// While we never check full difficulty transition logic, the mainnet option enables checking that
+/// difficulty transitions only happen every two weeks and never shift difficulty more than 4x in
+/// either direction, which is sufficient to prevent most minority hashrate attacks.
+///
+/// TODO: Update comment as headers are removed from cache when blocks are disconnected.
+/// We cache any headers which we connect until every block source is in agreement on the best tip.
+/// This prevents one block source from being able to orphan us on a fork of its own creation by
+/// not responding to requests for old headers on that fork. However, if one block source is
+/// unreachable this may result in our memory usage growing in accordance with the chain.
+pub struct MicroSPVClient<P: Poll> {
+	chain_tip: ValidatedBlockHeader,
+	chain_poller: P,
+	chain_notifier: ChainNotifier,
+}
+
+impl<P: Poll> MicroSPVClient<P> {
+
+	/// Creates a new `MicroSPVClient` with a chain poller for polling one or more block sources and
+	/// a chain listener for receiving updates of the new chain tip.
+	///
+	/// At least one of the polled `BlockSource`s must provide the necessary headers to disconnect
+	/// from the given `chain_tip` back to its common ancestor with the best chain assuming that its
+	/// height, hash, and chainwork are correct.
+	///
+	/// `backup_block_sources` are never queried unless we learned, via some `block_sources` source
+	/// that there exists a better, valid header chain but we failed to fetch the blocks. This is
+	/// useful when you have a block source which is more censorship-resistant than others but
+	/// which only provides headers. In this case, we can use such source(s) to learn of a censorship
+	/// attack without giving up privacy by querying a privacy-losing block sources.
+	pub fn init(chain_tip: ValidatedBlockHeader, chain_poller: P) -> Self {
+		let header_cache = HeaderCache::new();
+		let chain_notifier = ChainNotifier { header_cache };
+		Self { chain_tip, chain_poller, chain_notifier }
+	}
+
+	/// Check each source for a new best tip and update the chain listener accordingly.
+	/// Returns true if some blocks were [dis]connected, false otherwise.
+	pub async fn poll_best_tip<CL: ChainListener + Sized>(&mut self, chain_listener: &mut CL) ->
```

Comment:
Hmm, why move away from having the `MicroSPVClient` hold a reference to the `ChainListener`? It seems like it could make it easier to screw up tracking of best tip - I believe with the new API you'd need one `MicroSPVClient` per `ChainListener` (or we'd end up connecting blocks only on one listener), but the API doesn't make that obvious (naively, I'd implement the client-side by just calling poll on all the listeners in series). May also make sense to have a `add_sync_listener` which calls `init_sync_listener` as a part of adding a listener to the client.

Reply:
> Hmm, why move away from having the `MicroSPVClient` hold a reference to the `ChainListener`? It seems like it could make it easier to screw up tracking of best tip - I believe with the new API you'd need one `MicroSPVClient` per `ChainListener` (or we'd end up connecting blocks only on one listener), but the API doesn't make that obvious (naively, I'd implement the client-side by just calling poll on all the listeners in series).

Very good point. I was considering moving it back for that very reason and also because blocks would be needlessly re-fetched. Can't recall if I moved it to ease testing. Anyhow, I'll move it back.

> May also make sense to have a `add_sync_listener` which calls `init_sync_listener` as a part of adding a listener to the client.

This would imply the client holding multiple listeners vs the user implementing `ChainListener` as needed. For example, having in implementation calling out to both `ChannelManager` and `ChainMonitor`. Haven't put much thought into having a more flexible interface as this, but just wanted to call that out.

Reply:
> This would imply the client holding multiple listeners vs the user implementing ChainListener as needed. For example, having an implementation calling out to both ChannelManager and ChainMonitor. Haven't put much thought into having a more flexible interface as this, but just wanted to call that out.

Right, yea, either could work fine. In either case it may make sense to pull `init_sync_listener` into the `MicroSPVClient`, I think, though.

Reply:
> Right, yea, either could work fine. In either case it may make sense to pull `init_sync_listener` into the `MicroSPVClient`, I think, though.

I tried adding a member function that accepts a block source in a similar manner as `ChainPoller` (i.e., `<'b, B: DerefMut<Target=dyn BlockSource + 'b> + Sized + Sync + Send>`. However, attempting to use this with a block source already referenced by `ChainPoller` would result in two mutable borrows. I thought there was a workaround for this. Any advice here?

Reply:
Right, we'd need to (a) accept the `old_block` blockhash as an argument, (b) poll the poller for how to get to our current tip (but not *the* current tip, it cant change between sources) from `old_block`, and then (c) somehow track the sources that are done as a part of `MicroSPVClient`.

Reply:
So that requires a bit of refactoring but is doable. The problem I'm having is that -- when writing a the test -- I need to initialize `MicroSPVClient` with a `ChainPoller` that has a mutable reference to a block source and then try to pass the same block source to the new method to do the sync. But that leads to a compilation error because there are two mutable references. I thought there may be some trick with `Arc` / `Mutex` to work around this but couldn't figure it out.

Reply:
Right, I figured that you wouldn't need to pass the block source to the new `init_sync_listener` - it would just make some request through the `ChainPoller` to get a block source/block data.

Reply:
Yeah, that's easier, but it would mean possibly fetching through multiple sources if a different implementation of `Poll` is used.

Reply:
I think we're maybe out of sync here, but no matter - in general using different sources shouldn't matter - they are all "trusted" to give us the right transaction data, presumably. As long as we can (a) iteratively add new listeners and sync them up to the same block in a safe way that makes accidentally syncing them to different blocks hard, and (b) then let the spv client run, keeping all the listeners up to date, we should be good.

Reply:
Gotcha. I was referring to the comment above `init_sync_listener`:

https://github.com/rust-bitcoin/rust-lightning/blob/eaedc0a8a3cce989aa792daff3737b48201dec23/lightning-block-sync/src/lib.rs#L390-L394

Reply:
Right, I think that comment was just a general note about how that block source could leave us un-sync-able if it drove us onto some fork and then refused to provide previous headers on the chain to the main chain. Yet another reason to move it into the rest of the spvclient logic - we could use the disconnected-headers caching to ensure that no single block source could do that to us.